src/java.base/share/classes/java/lang/invoke/X-VarHandleMemoryAccess.java.template

```diff
@@ -115,14 +115,15 @@ final class MemoryAccessVarHandle$Type$Helper extends MemoryAccessVarHandleBase
     static long offsetNoVMAlignCheck(boolean skipAlignmentMaskCheck, MemorySegmentProxy bb, long offset, long alignmentMask) {
         long base = bb.unsafeGetOffset();
         long address = base + offset;
-        if (skipAlignmentMaskCheck) {
+        long maxAlignMask = bb.maxAlignMask();
+        if (skipAlignmentMaskCheck && maxAlignMask == 0) {
```

Comment:
I've got minor comment. I wonder if we can `skipAlignmentMaskCheck` parameter and just base on `maxAlignMask`

Reply:
Sorry, it was stupid question. Please ignore

Reply:
No problem - this is quite tricky. In general, `skipAlignmentMaskCheck` still does _something_ when accessing native segments (as the max align mask is zero there), but in general you are morally correct in observing that it's less useful than it used to be. My hope is that with the right VM fixes, we can get rid of `skipAlignmentMaskCheck` completely, and simplify all this.

For the records, the difference between unaligned and aligned access is as follows:

```
Benchmark                                                Mode  Cnt  Score   Error  Units
LoopOverNonConstant.segment_loop_instance                avgt   30  0.229 ? 0.002  ms/op
LoopOverNonConstant.segment_loop_instance_aligned        avgt   30  0.335 ? 0.001  ms/op
```

It's not terrible, but could of course be better.

Reply:
Hmm...

I wonder a bit if this code can be optimized `if (((address | maxAlignMask) & alignmentMask) != 0) {`

Max `maxAlignMask`, can be 1,2,4 (if I see it correctly) so if `maxAlignMask` `alignmentMask` and will not be correct, than exception will be thrown regardless of address? I wonder if making static check could be more performant as VM move such check before loop, than we could only focus on `address & alignmentMask != 0`. Or something like this.

Reply:
The max align check is already moved out of the loop (as it only depends on the segment, which is a loop invariant). The problematic check, which is not moved outside the loop is exactly `address & alignmentMask != 0`. That check cannot be hoisted, because `address` is made up of `offset` and the segment base address; the latter is a loop invariant, but the former isn't.

Of course C2 should be able to detect that offset is a function of the loop variable, and prove that the check can never fail - but this pattern is currently not detected (hence JDK-8277850).

Reply:
Yes, I was thinking about line 125 `if (((address | maxAlignMask) & alignmentMask) != 0) {` to split into to two
`if ((maxAlignMask & alignmentMask) != 0) throw...` and `if (address & alignmentMask) != 0) throw...` if this could help a bit.